tree-walk.c

```diff
@@ -600,9 +630,20 @@ enum follow_symlinks_result get_tree_entry_follow_symlinks(struct object_id *tre
 			void *tree;
 			struct object_id root;
 			unsigned long size;
-			tree = read_object_with_reference(&current_tree_oid,
-							  tree_type, &size,
-							  &root);
+			if (gently) {
+				struct object_info oi = OBJECT_INFO_INIT;
+				if (oid_object_info_extended(the_repository,
+				    &current_tree_oid, &oi, 0) < 0) {
+					retval = MISSING_OBJECT;
+					goto done;
+				}
+				tree = *oi.contentp;
+			} else {
+				tree = read_object_with_reference(&current_tree_oid,
+								  tree_type, &size,
+								  &root);
+			}
```

Comment:
Once again: simpler conditional block first.

Reply:
(I'll not repeat this for the rest of the file, it really applies to all if...else conditionals ;-))